File: solutions/173.binary-search-tree-iterator.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
class BSTIterator
  attr_accessor :stack

  #   :type root: TreeNode
  def initialize(root)
    @stack = []
    until root.nil?
      @stack << root
      root = root.left
    end
  end

  #   :rtype: Integer
  def next
    node = stack.pop
    val = node.val
    node = node.right
    until node.nil?
      stack << node
      node = node.left
    end
    val
  end

  #   :rtype: Boolean
  def has_next
    !stack.empty?
  end
end

# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator.new(root)
# param_1 = obj.next()
# param_2 = obj.has_next()

File: solutions/173.binary-search-tree-iterator.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from typing import Optional


class BSTIterator:

    def __init__(self, root: Optional[TreeNode]):
        self.stack = []
        while root:
            self.stack.append(root)
            root = root.left

    def next(self) -> int:
        node = self.stack.pop()
        val = node.val
        node = node.right
        while node:
            self.stack.append(node)
            node = node.left
        return val

    def hasNext(self) -> bool:
        return bool(self.stack)


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()
